crates/ruff/src/rules/ruff/rules/mutable_defaults_in_dataclass_fields.rs

```diff
@@ -0,0 +1,207 @@
+/// ```
+#[violation]
+pub struct FunctionCallInDataclassDefaultArgument {
+    pub name: Option<String>,
+}
+
+impl Violation for FunctionCallInDataclassDefaultArgument {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let FunctionCallInDataclassDefaultArgument { name } = self;
+        if let Some(name) = name {
+            format!("Do not perform function call `{name}` in dataclass defaults")
+        } else {
+            format!("Do not perform function call in dataclass defaults")
+        }
+    }
+}
+
+fn is_mutable_expr(expr: &Expr) -> bool {
+    matches!(
+        &expr.node,
+        ExprKind::List { .. }
+            | ExprKind::Dict { .. }
+            | ExprKind::Set { .. }
+            | ExprKind::ListComp { .. }
+            | ExprKind::DictComp { .. }
+            | ExprKind::SetComp { .. }
+    )
+}
+
+const ALLOWED_FUNCS: &[&[&str]] = &[&["dataclasses", "field"]];
+
+fn is_allowed_func(checker: &Checker, func: &Expr) -> bool {
+    checker
+        .ctx
+        .resolve_call_path(func)
+        .map_or(false, |call_path| {
+            ALLOWED_FUNCS
+                .iter()
+                .any(|target| call_path.as_slice() == *target)
+        })
+}
+
+pub fn function_call_in_dataclass_defaults(checker: &mut Checker, body: &[Stmt]) {
+    for statement in body {
+        if let StmtKind::AnnAssign {
+            value: Some(expr), ..
+        } = &statement.node
+        {
+            if let ExprKind::Call { func, .. } = &expr.node {
+                if !is_allowed_func(checker, func) {
+                    checker.diagnostics.push(Diagnostic::new(
+                        FunctionCallInDataclassDefaultArgument {
+                            name: compose_call_path(func),
+                        },
+                        Range::from(expr),
+                    ));
+                }
+            }
+        }
+    }
+}
+
+pub fn mutable_dataclass_default(checker: &mut Checker, body: &[Stmt]) {
+    for statement in body {
+        match &statement.node {
+            StmtKind::AnnAssign {
+                value: Some(value), ..
+            }
+            | StmtKind::Assign { value, .. } => {
+                if is_mutable_expr(value) {
+                    checker
+                        .diagnostics
+                        .push(Diagnostic::new(MutableDataclassDefault, Range::from(value)));
+                }
+            }
+            _ => (),
+        }
+    }
+}
+
+pub fn is_dataclass(checker: &Checker, decorator_list: &[Expr]) -> bool {
+    for decorator in decorator_list.iter() {
+        if checker
+            .ctx
+            .resolve_call_path(if let ExprKind::Call { func, .. } = &decorator.node {
+                func
+            } else {
+                decorator
+            })
+            .map_or(false, |call_path| {
+                call_path.as_slice() == ["dataclasses", "dataclass"]
+            })
+        {
+            return true;
+        }
+    }
+    false
+}
```

Comment:
Unless I'm misreading, I think you can write this as:

```rust
pub fn is_dataclass(checker: &Checker, decorator_list: &[Expr]) -> bool {
    decorator_list.iter().any(|decorator| {
        checker
            .ctx
            .resolve_call_path(map_callable(decorator))
            .map_or(false, |call_path| {
                call_path.as_slice() == ["dataclasses", "dataclass"]
            })
    })
}
```

Two things: (1) `decorator_list.iter().any` over manual iteration, and (2) we have a `map_callable` utility to handle the `if let ExprKind::Call { ... }` pattern here since it comes up often with decorators.

Reply:
I knew something like that existed, thank you! Did replace both.